packages/server-wallet/src/wallet/__test__/integration/push-message.test.ts

```diff
@@ -359,3 +372,245 @@ describe('when there is a request provided', () => {
     });
   });
 });
+
+describe('ledger funded app scenarios', () => {
+  let ledger: Channel;
+  let app: Channel;
+  let expectedUpdatedLedgerState: SignedStateWithHash;
+
+  const preFS = {
+    turnNum: 0,
+    outcome: simpleEthAllocation([{destination: aliceP().destination, amount: BN.from(5)}]),
+  };
+
+  beforeEach(async () => {
+    // FIXME: SQL-ize ledgers and do test cleanup as needed
+    wallet.store.ledgers = {};
+    wallet.store.ledgerRequests = new LedgerRequests();
+
+    const someNonConflictingChannelNonce = 23364518;
+
+    // NOTE: Put a ledger Channel in the DB
+    ledger = await Channel.query(wallet.knex).insert(
+      channel({
+        channelNonce: someNonConflictingChannelNonce,
+        vars: [
+          stateWithHashSignedBy(
+            alice(),
+            bob()
+          )({
+            appDefinition: '0x0000000000000000000000000000000000000000',
+            channelNonce: someNonConflictingChannelNonce,
+            turnNum: 4,
+            outcome: simpleEthAllocation([{destination: aliceP().destination, amount: BN.from(5)}]),
+          }),
+        ],
+      })
+    );
+
+    // Update the in-memory Ledgers table
+    wallet.__setLedger(ledger.channelId, ETH_ASSET_HOLDER_ADDRESS);
+
+    // Generate application channel
+    app = channel({
+      fundingStrategy: 'Ledger',
+    });
+
+    // Construct expected ledger update state
+    expectedUpdatedLedgerState = {
+      // eslint-disable-next-line
+      ...ledger.latest!,
+      turnNum: 6,
+      outcome: {
+        type: 'SimpleAllocation' as const,
+        // TODO: Avoid hardcoded response
+        assetHolderAddress: '0x0000000000000000000000000000000000000000',
+        allocationItems: [
+          {
+            destination: makeDestination(app.channelId), // Funds allocated to channel
+            amount: BN.from(5), // As per channel outcome
+          },
+        ],
+      },
+    };
+  });
+
+  const putTestChannelInsideWallet = async (args: PartialModelObject<Channel>) => {
+    const channel = await Channel.query(wallet.knex).insert(args);
+
+    // Add the objective into the wallets store (normally would have happened
+    // during createChannel or pushMessage call by the wallet)
+    await ObjectiveModel.insert(
+      {
+        type: 'OpenChannel',
+        status: 'approved',
+        participants: channel.participants,
+        data: {
+          targetChannelId: channel.channelId,
+          fundingStrategy: 'Unfunded', // Could also be Direct, funding is empty
+        },
+      },
+      wallet.knex
+    );
+
+    return channel;
+  };
+
+  it('countersigns a prefund setup and automatically creates a ledger update', async () => {
```

Comment:
More of a meta-comment, but it seems bad that the `pushMessage` test has so many disparate test cases.